climakitae/threshold_tools.py

```diff
@@ -0,0 +1,613 @@
+import numpy as np
+import pandas as pd
+import xarray as xr
+from scipy import stats
+
+import lmoments3 as lm
+from lmoments3 import distr as ldistr
+from lmoments3 import stats as lstats
+
+import matplotlib.pyplot as plt
+from matplotlib import cm
+import cartopy.crs as ccrs
+import cartopy.feature as cfeature
+import geoviews as gv
+
+import holoviews as hv
+from holoviews import opts
+import hvplot.pandas
+import hvplot.xarray
+
+from .visualize import get_frequency_plot, get_geospatial_plot
+
+#####################################################################
+
+
+def get_ams(da, extremes_type="max"):
+
+    extremes_types = ["max"]
+    if extremes_type not in extremes_types:
+        raise ValueError(
+            "invalid extremes type. expected one of the following: %s" % extremes_types
+        )
+
+    if extremes_type == "max":
+        ams = da.resample(time="A").max(keep_attrs=True)
+        ams.attrs["extreme value extraction method"] = "block maxima"
+        ams.attrs["extremes type"] = "maxima"
+        ams.attrs["block size"] = "1 year"
+        ams.attrs["timeseries type"] = "annual max series"
+
+    # if extremes_type == 'min':
+    #     ams = da.resample(time='A').min(keep_attrs=True)
+    #     ams.attrs['extreme value extraction method'] = 'block maxima'
+    #     ams.attrs['extremes type'] = 'minima'
+    #     ams.attrs['block size'] = '1 year'
+    #     ams.attrs['timeseries type'] = 'annual min series'
+
+    return ams
+
+
+#####################################################################
+
+
+def get_lmoments(ams, distr="gev", multiple_points=True):
+
+    distrs = ["gev", "gumbel", "weibull", "pearson3", "genpareto"]
+    if distr not in distrs:
+        raise ValueError(
+            "invalid distr type. expected one of the following: %s" % distrs
+        )
+
+    ams_attributes = ams.attrs
+
+    if multiple_points == True:
+        ams = ams.stack(allpoints=["x", "y"]).squeeze().groupby("allpoints")
+
+    def apply_function(ams):
+        def function(ams):
+
+            if distr == "gev":
+                lmoments = ldistr.gev.lmom_fit(ams)
+
+            if distr == "gumbel":
+                lmoments = ldistr.gum.lmom_fit(ams)
+
+            if distr == "weibull":
+                lmoments = ldistr.wei.lmom_fit(ams)
+
+            if distr == "pearson3":
+                lmoments = ldistr.pe3.lmom_fit(ams)
+
+            if distr == "genpareto":
+                lmoments = ldistr.gpa.lmom_fit(ams)
+
+            return lmoments
+
```

Comment:
time dimension is removed because multiple years (ex. 30 years of extreme value points) are inputted to produce an output that is just a single return value or probability. if time isn't removed, there's an error that occurs.